src/targets/gpu/fuse_ck.cpp

```diff
@@ -161,10 +203,81 @@ struct find_ck_gemm
     }
 };
 
+static bool is_mul_module(const module& m)
+{
+    std::vector<std::string> result;
+    for(auto& ins : m)
+    {
+        if(starts_with(ins.name(), "@"))
+            continue;
+        if(contains({"multibroadcast", "contiguous"}, ins.name()))
+            continue;
+        if(ins.name() == "pointwise")
+        {
+            return is_mul_module(*ins.module_inputs().front());
+        }
+        else if(ins.name() == "mul")
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+struct find_ck_gemm_softmax_gemm
+{
+    auto matcher() const
+    {
+        auto gemm1 =
+            match::skip(match::name("contiguous"))(match::name("dot")(is_ck_gemm().bind("gemm1")));
+        auto mul = match::name("pointwise")(match::any_of[match::inputs()](gemm1)).bind("scale");
+        auto softmax = match::name("softmax")(match::any_of[match::inputs()](mul)).bind("softmax");
+        return match::name("dot")(is_ck_gemm().bind("gemm2"))(
+            match::any_of[match::inputs()](softmax));
+    }
+
+    void apply(module_pass_manager& mpm, const match::matcher_result& r) const
+    {
+        auto ins       = r.result;
+        auto gemm2_ins = r.instructions["gemm2"];
+        auto gemm1_ins = r.instructions["gemm1"];
+        auto scale_ins = r.instructions["scale"];
+
+        if(scale_ins->module_inputs().size() != 1 or
+           not is_mul_module(*scale_ins->module_inputs().front()))
+            return;
+        if(not ck_gemm_softmax_gemm::is_ck_supported_type(gemm1_ins->get_shape().type()))
+            return;
+
+        double scale = 1.0;
+        for(auto& in : scale_ins->inputs())
+        {
+            if(in->can_eval())
+            {
+                in->get_literal().visit([&](const auto s) {
+                    if(std::all_of(s.begin() + 1, s.end(), [&](auto v) {
+                           return float_equal(v, s.front());
+                       }))
+                        scale = s.front();
+                    else
+                        return;
+                });
+            }
+        }
+
+        auto inputs = gemm1_ins->inputs();            // A, B
+        inputs.push_back(gemm2_ins->inputs().back()); // B1
+
+        mpm.get_module().replace_instruction(
+            ins, ck_gemm_softmax_gemm{gemm2_ins->get_operator(), scale}, inputs);
+    }
+};
+
 } // namespace
 
 void fuse_ck::apply(module_pass_manager& mpm) const
 {
+    match::find_matches(mpm, find_ck_gemm_softmax_gemm{});
```

Comment:
I dont think we need an extra `find_matches`, I think this matcher could be added below.

Reply:
We need to match in the order of gemm-softmax-gemm, gemm-pointwise, gemm. When I put the matchers in the same `find_matches` call, I didn't see that order being maintained regardless of the order in which they were passed.

Reply:
I was thinking `match::find_matches(mpm, find_ck_gemm_softmax_gemm{}, find_ck_gemm_pointwise{})`, does that work?

Reply:
Both `match::find_matches(mpm, find_ck_gemm_softmax_gemm{}, find_ck_gemm_pointwise{});` and `match::find_matches(mpm, find_ck_gemm_pointwise{}, find_ck_gemm_softmax_gemm{});` seem to apply `find_ck_gemm_pointwise` first, which prevents `find_ck_gemm_softmax_gemm` from working.

Reply:
I see, thats because it matches the last gemm.